File: Dijkstra.go
package main

import (
	"fmt"
	"math"
)

const INF = math.MaxInt32

func dijkstra(graph [][]int, start int) []int {
	n := len(graph)
	dist := make([]int, n)
	visited := make([]bool, n)

	// Initialisation des distances
	for i := 0; i < n; i++ {
		dist[i] = INF
	}
	dist[start] = 0

	for i := 0; i < n-1; i++ {
		u := -1
		min := INF

		// Trouver le sommet non visité avec la plus petite distance
		for j := 0; j < n; j++ {
			if !visited[j] && dist[j] < min {
				min = dist[j]
				u = j
			}
		}

		if u == -1 {
			break
		}

		visited[u] = true

		// Mise à jour des distances
		for v := 0; v < n; v++ {
			if !visited[v] && graph[u][v] != 0 &&
				dist[u] != INF &&
				dist[u]+graph[u][v] < dist[v] {
				dist[v] = dist[u] + graph[u][v]
			}
		}
	}

	return dist
}

func main() {
	graph := [][]int{
		{0, 10, 0, 0, 5},
		{10, 0, 1, 0, 2},
		{0, 1, 0, 4, 0},
		{0, 0, 4, 0, 3},
		{5, 2, 0, 3, 0},
	}

	start := 0
	distances := dijkstra(graph, start)

	fmt.Println("Distances depuis le sommet", start)
	for i, d := range distances {
		fmt.Printf("Vers %d : %d\n", i, d)
	}
}

File: Dijkstra.rs
use std::i32;

const INF: i32 = i32::MAX;

fn dijkstra(graph: Vec<Vec<i32>>, start: usize) -> Vec<i32> {
    let n = graph.len();
    let mut dist = vec![INF; n];
    let mut visited = vec![false; n];

    dist[start] = 0;

    for _ in 0..n - 1 {
        let mut min = INF;
        let mut u: Option<usize> = None;

        // Trouver le sommet non visité avec la distance minimale
        for i in 0..n {
            if !visited[i] && dist[i] < min {
                min = dist[i];
                u = Some(i);
            }
        }

        let u = match u {
            Some(v) => v,
            None => break,
        };

        visited[u] = true;

        // Mettre à jour les distances des voisins
        for v in 0..n {
            if !visited[v]
                && graph[u][v] != 0
                && dist[u] != INF
                && dist[u] + graph[u][v] < dist[v]
            {
                dist[v] = dist[u] + graph[u][v];
            }
        }
    }

    dist
}

fn main() {
    let graph = vec![
        vec![0, 7, 0, 0, 3],
        vec![7, 0, 1, 2, 0],
        vec![0, 1, 0, 6, 0],
        vec![0, 2, 6, 0, 4],
        vec![3, 0, 0, 4, 0],
    ];

    let start = 0;
    let distances = dijkstra(graph, start);

    println!("Distances depuis le sommet {}", start);
    for (i, d) in distances.iter().enumerate() {
        println!("Vers {} : {}", i, d);
    }
}
